### PR TITLE
Port Sponge Async lighting patch

### PR DESCRIPTION
**Pull request by mikroskeem**
I did quick copypasta from my [initial try](https://github.com/mikroskeem/asynclighting) of async patch port.

~~Known needed changes:~~
- ~~Add configuration option for async thread amount~~
- ~~Triple check the correct mappings (as code comes from MCP and... CB mappings are the best! /s)~~
- ~~Remove bunch of `// MCP - ...` comments~~
- ~~More OBFHELPERS (and thus imports) needed!~~
- ~~Paper start/end comments need to be improved (imo)~~

~~Do not merge yet, code needs review and changes listed above (atleast)~~

Ready for proper review now.

Edit: Do not merge, code is broken for unknown reasons right now. Will take a look soon what I did wrong.

### CODE REVIEW

File: Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+ 
++    public void setSkyLight(int i, int j, int k, int l) { a(i, j, k, l); } // Paper - OBFHELPER
+     public void a(int i, int j, int k, int l) {
+         this.skyLight.a(i, j, k, l);
+     }
+@@ -95,6 +96,7 @@ public class ChunkSection {
+         return this.skyLight.a(i, j, k);
+     }
+ 
++    public void setBlockLight(int i, int j, int k, int l) { b(i, j, k, l); }
```
Comment (aikar) [inline Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch]: Missing // Paper - OBFHELPER

File: Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch [hunk 2]
```diff
@@ -0,0 +1,1221 @@
...
+ 
++    public int getTopFilledSegment() { return g(); } // Paper - OBFHELPER
+     public int g() {
+         ChunkSection chunksection = this.y();
+ 
+@@ -274,6 +290,9 @@ public class Chunk {
+     }
+ 
+     private void h(boolean flag) {
++        // Paper start - asynclighting
```
Comment (aikar) [inline Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch]: Is it intended that any calls to the previous recheckGaps is now no longer going to do anything....?

File: Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch [hunk 3]
```diff
@@ -0,0 +1,1221 @@
...
+         this.sections = new ChunkSection[16];
+         this.g = new byte[256];
+         this.h = new int[256];
+@@ -151,6 +152,18 @@ public class Chunk {
+     public boolean mustSave;
+     // CraftBukkit end
+ 
++    // Paper start - asynclighting
++    // Keeps track of block positions in this chunk currently queued for sky light update
++    private java.util.concurrent.CopyOnWriteArrayList<Short> queuedSkyLightingUpdates = new java.util.concurrent.CopyOnWriteArrayList<>();
```
Comment (aikar) [inline Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch]: Can we use a ConcurrentLinkedQueue for all of these?

COWAL imo should not be used for things you expect lots of changes to.

File: Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch [hunk 4]
```diff
@@ -0,0 +1,1221 @@
...
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -276,4 +276,9 @@ public class PaperConfig {
+     private static void authenticationServersDownKickMessage() {
+         authenticationServersDownKickMessage = Strings.emptyToNull(getString("messages.kick.authentication-servers-down", authenticationServersDownKickMessage));
+     }
++
++    public static int asyncLightingThreads = 2;
++    private static void asyncLightingThreads() {
++        asyncLightingThreads = getInt("settings.async-lighting-threads", 2);
```
Comment (aikar) [inline Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch]: I'm actually in favor of not making this configurable at all.

It's going to lead to people thinking more threads = better.

**Review by aikar — CHANGES_REQUESTED** [review]
As mentioned on IRC. I'd like to see most of the sponge neighbor changes removed, and keep to the CB system as they seem to function the same.

However, we can keep just the new additions to maintain the new getNeighbors()/getNeighbor() method to get direct chunk refs, as that avoids map lookups.

So can you remove all the areNeighborsLoaded changes and keep that logic to as it was and only keep the parts to maintain the neighbors chunk array.

**Comment by daboluo666** [comment]
Awesome! I believe this has a good performance improvement, what is hindering your plan? @mikroskeem

**Comment by electronicboy** [comment]
There is currently an issue where the patch basically ends up breaking world generation, I have some levels of theory into what is going on, but need to look closer at CBs/Spigots changes in regards to ticking chunks

**Comment by SecretlyJealous** [comment]
Really hoping this patch is sorted before 1.13 comes out, been stuck using Paper-1228 because of the lag/crashes that occur when calculating the lightmap properly on future builds.

**Comment by electronicboy** [comment]
In the latest build (https://ci.destroystokyo.com/job/PaperSpigot/1275/), I have brought back spigots broken lighting logic, it was never intended that we'd leave this change in without the async chunk lighting for this long, however, shamefully I haven't had the time to sit down and work out the issues around craftbukkits chunk ticking logic which is causing breakage.

**Comment by kashike** [comment]
@mikroskeem Do you plan on completing this?

**Comment by zachbr** [comment]
@kashike I believe @electronicboy has already pulled this into a feature branch and was doing work on it there. My understanding is that the changes in that branch are more up-to-date than the ones in this PR. https://github.com/PaperMC/Paper/tree/feature/async-lighting

**Comment by mikroskeem** [comment]
In that case yes, this can be closed.

**Pull request closed without merging**